Let user define search root for packages

RunsFor opened the pull request:
**Intro**:

By default, Lua uses package.path/cpath to load dependencies when user
calls `require("something")`. It also allows to add custom loaders to
extend its functionality. Tarantool transforms default behavior by
injecting some custom loaders into `package.loaders`.

Firstly, it allows to load modules from CWD. Secondly, it looks for
rocks modules starting from CWD and traversing up to root "/"
directory. This "start from CWD" behaviour is hardcoded and not
customizable.

**Problem**:

Let say we have a lua app with all dependencies under myapp directory.
If you start your application like this, you won't be able to load any
of them unless you specify myapp in every require call:

$ tarantool myapp/init.lua

To allow user scripts require dependencies relatively, user has to first
determine appropriate directory path relatively to scripts file and then
patch package.path/cpath with both modules and rocks paths.

**Solution**:

Introduce the notion of a search root which is used as base for rocks
and module loaders. Let user easily set this path to control how
application dependencies are loaded.

What this patch does:

`package.setsearchroot(path)` function sets search root which is used
as root directory to load dependencies from:

- path must be a string (relative or absolute). It will be expanded to
  absolute path and set as a search root;
- if no path specidied, it sets current file directory as a search root
  (using debug.sourcedir);
- if path is box.NULL, that searchroot will be reset,
  so default behaviour will apply (CWD).

With that said everything a user now has to do is to just put this line
of code:

```lua
package.setsearchroot()
```

at the top of project init file "myapp/init.lua" which is placed in the
project root and start an app by simply calling:

```bash
$ tarantool myapp/init.lua
```

This will set search root to the absolute path of `myapp` directory, so
all dependencies will be looked relative to that directory.

rosik commented on src/lua/init.lua:
```diff
@@ -194,15 +227,19 @@ local function search(name)
+table.insert(package.loaders, 4, get_loader_func({path_fn = approot, is_lib = false}))
+table.insert(package.loaders, 5, get_loader_func({path_fn = approot, is_lib = true}))
+table.insert(package.loaders, 6, get_loader_func({path_fn = get_rocksluadir, is_lib = false, traverse = true}))
+table.insert(package.loaders, 7, get_loader_func({path_fn = get_rockslibdir, is_lib = true, traverse = true}))
```
This bunch looks strange.

1. `is_lib` option partially overlaps what is passed to `path_fn`.
2. why does it accept function as an argument?
3. when approot is not set loaders are duplicated

RunsFor replied:
> 1. is_lib option partially overlaps what is passed to path_fn.

Sorry, I didn't get the part about overlapping. `is_lib` flag lets loader function to define both searchpath templates (i.e. `?.so` for libs) and the way how this module will be loaded (i.e. using `package.loadlib` for libs). `path_fn` is a simple function, which returns some path which will be later used to append lib or lua template.

> 2. why does it accept function as an argument?

function for paths are needed to be able to get an actual path at runtime. I.e. current working directory may be changed during runtime, so we need to get the right one when doing `require`

> 3. when approot is not set loaders are duplicated

You are right. But I don't know how to improve it right now. We need to add this loader anyway.

RunsFor replied:
Refactored `get_loader_func` api to eliminate `is_lib` parameter. Now it looks much more readable

rosik commented on src/lua/init.lua:
```diff
@@ -86,77 +114,85 @@ local function load_lua(file)
+    end
+
+    return paths
 end
```
How about calling `fio.dirname` in a loop instead of using regex hacks?

rosik replied:
Also maybe it should be renamed to `traverse_path`?

RunsFor replied:
I agree. I switched to `fio.dirname` and `traverse_path` seems more appropriate name here.

rosik commented on src/lua/init.lua:
```diff
@@ -86,77 +114,85 @@ local function load_lua(file)
-    local search_cwd = lib and search_cwd_lib or search_cwd_lua
-    local load_func = lib and load_lib or load_lua
+local function traverse_path(path)
+    path = path:sub(1, 1) == '/' and path or fio.abspath(path)
```
```console
> traverse_path('/a/b/c/../../A')
---
- - /a/b/c/../../A
  - /a/b/c/../..
  - /a/b/c/..
  - /a/b/c
  - /a/b
  - /a
  - /
...
```

Is this appropriate behavior? Maybe `fio.abspath` should be called always?

RunsFor replied:
This is how it works right now. Your suggestion is reasonable and I agree with it, but it is a changing behaviour.

rosik replied:
Ok, it's reasonable to keep it then.

rosik commented on src/lua/init.lua:
```diff
@@ -86,77 +114,85 @@ local function load_lua(file)
+
+    table.insert(paths, path)
+
+    return paths
```
```lua
    local paths = {path}
    while path ~= '/' do
        path = fio.dirname(path)
        table.insert(paths, path)
    end
    return paths
```

RunsFor replied:
✅

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
-    local search_rocks = lib and search_rocks_lib or search_rocks_lua
-    local load_func = lib and load_lib or load_lua
-    return function (name)
+local function get_loader_func(opts)
```
I would name it `gen_loader_func` (and the same for `get_search_func` → `gen_search_func`).

RunsFor replied:
Renamed

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+local function get_search_func(opts)
+    local path_fn = opts.path_fn
+    if type(opts.path_fn) ~= 'function' then
+        path_fn = function() return opts.path_fn end
```
Dead code.

RunsFor replied:
Fixed

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
-        else
-            return err
-        end
+        local path = path_fn() or '.'
```
`or '.'` is needed only to use cwd when approot is not set? Why not to skip this loader in the case at all? Otherwise you'll search in cwd twice (for regular modules, not rocks).

RunsFor replied:
How do you skip a loader in Lua?

Totktonada replied:
Return empty string.

Anyway, we discussed this voicely with you and decided to 'merge' cwd and appdir loaders (use appdir when available, otherwise cwd), so skipping is not needed.

RunsFor replied:
Removed possible duplication of loaders, but let `path_fn() or '.'` stay for the say of consistency

Totktonada replied:
Now searchroot can return `nil`, so `or '.'` is needed. This however the only case when we pass a relative path into `traverse_path()`, so I would use `or fio.cwd()` here. We however should leave `fio.abspath()` in `traverse_path()` to suppress `/../` parts.

Totktonada replied:
Ouch, no, `/../` parts are suppressed in `setsearchroot()`.

Totktonada replied:
Now the only style question `or '.'` vs `or fio.cwd()` is actual here. It is not so important. Marked as resolved.

Totktonada commented on src/lua/init.lua:
```diff
@@ -44,6 +44,26 @@ tarantool_exit(int);
 
 local fio = require("fio")
 
+local function sourcepath()
```
I think `path` is unclear in whether it is a path to a file or to a directory.

RunsFor replied:
Did it in #4260

Totktonada commented on src/lua/init.lua:
```diff
@@ -44,6 +44,26 @@ tarantool_exit(int);
+
+local package_approot
+
+local function approot(path)
```
Maybe name it `search_root`?

RunsFor replied:
Made it `searchroot`

Totktonada commented on src/lua/init.lua:
```diff
@@ -194,15 +211,31 @@ local function search(name)
+    search_fn = get_search_func({ path_fn = approot, templates = ROCKS_LUA_TEMPLATES, traverse = true }),
+    load_fn = load_lua }))
+table.insert(package.loaders, 7, get_loader_func({
+    search_fn = get_search_func({ path_fn = approot, templates = ROCKS_LIB_TEMPLATES, traverse = true }),
```
We try to load packages from cwd and approot both, but rocks only from approot. Don't sure whether it is right, but looks inconsistent. Say, we cannot state that setting an approot directory don't break cwd-relative requires.

RunsFor replied:
Fixed it. See comment above

Totktonada commented on src/lua/init.lua:
```diff
@@ -194,15 +211,31 @@ local function search(name)
+    load_fn = load_lua }))
+table.insert(package.loaders, 3, get_loader_func({
+    search_fn = get_search_func({ path_fn = fio.cwd, templates = LIB_TEMPLATES }),
+    load_fn = load_lib }))
```
Rocks are searched in approot first (if set), but regular packages are searched first in cwd. Looks inconsistent.

RunsFor replied:
Made it more consistent by merge cwd loader with searchroot loader. Now if no searchroot defined, it acts as simple cwd loader

Totktonada commented:
@RunsFor BTW, it seems that you close #3134 in this PR.

Totktonada commented on src/lua/init.lua:
```diff
@@ -44,6 +44,26 @@ tarantool_exit(int);
 local fio = require("fio")
 
+local function sourcepath()
+    local sourcepath = debug.getinfo(2, "S").source:sub(2):match("(.*/)") or './'
```
Raw idea: can we check stack frames count to either adjust a frame number or at least give an error about inappropriate use of the function?

akudiyar replied:
Just an idea: why don't write the source path and other file information directly into `package.loaded` by the loader? Or we need the source path information earlier?

Totktonada replied:
I think it is good to have some flexibility here. Say, I need to set `${PWD}/../..` as a directory to require graphql implementation from a graphql test.

RunsFor replied:
It was simplified a bit and moved to the upstream in #4260.

rosik commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+end
+
+local function traverse_path(path)
+    path = path:sub(1, 1) == '/' and path or fio.abspath(path)
```
I've already pointed on always calling abspath, but it changes behavior. We can discuss it once again.

Totktonada replied:
Sorry, missed your comment. This should be commented at least, it is not obvoius point.

RunsFor replied:
This is undocumented `feature`. Maybe we should just fix it and mention about it in changelog. @Totktonada do you mind?

Totktonada replied:
I'll cite Yaroslav's example to bring needed context:

```lua
> traverse_path('/a/b/c/../../A')
---
- - /a/b/c/../../A
  - /a/b/c/../..
  - /a/b/c/..
  - /a/b/c
  - /a/b
  - /a
  - /
...
```

A path with `/../` in a middle could not appear in `traverse_rocks()`, because `fio.cwd()` does not produce such paths. So there is no need to introduce such strange behaviour and so it worth to always call `fio.abspath()`. Ilya, do not I missed something?

Totktonada replied:
Ouch, no, `fio.abspath()` doing this work in `setsearchroot()`, so there is no need in `fio.abspath()` here at all.

RunsFor replied:
`traverse_path` function now uses `fio.abspath`

RunsFor commented on src/lua/init.lua:
```diff
@@ -169,20 +189,17 @@ local function rocks_loader_func(lib)
+        get_search_func({path_fn = fio.cwd, templates = ROCKS_LUA_TEMPLATES, traverse = true}),
+        get_search_func({path_fn = fio.cwd, templates = ROCKS_LIB_TEMPLATES, traverse = true}),
+        get_search_func({path_fn = function() return package.path end}),
+        get_search_func({path_fn = function() return package.cpath end}),
```
Adjusted code a little bit. I made it more consistent with `require`, so it now based on what we call search root. But I was not able to fully reuse code. Anyway it is much more readable now

Totktonada replied:
Reworked the code to avoid the hack with `path_fn()` with package.{path,cpath} and duplication of the search function definitions.

RunsFor commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+local function get_loader_func(opts)
+    opts = opts or {}
+    assert(type(opts.load_fn) == 'function', 'load_fn must be defined')
+    opts.search_fn = opts.search_fn or function() return fio.cwd end
```
Fixed.

RunsFor commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
-        "/.rocks/lib/tarantool/?."..soext
-    }
-    return traverse_rocks(name, pathes_search)
+local function search_lua(name, path)
```
Removed

RunsFor commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
+
+local function search_lib(name, path)
```
Removed

RunsFor commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
+local LIB_TEMPLATES = { '?.'..soext }
+local LUA_TEMPLATES = { '?.lua', '?/init.lua' }
+local ROCKS_LIB_TEMPLATES = { ROCKS_LIB_PATH .. '/?.'..soext }
+local ROCKS_LUA_TEMPLATES = { ROCKS_LUA_PATH .. '/?.lua', ROCKS_LUA_PATH .. '/?/init.lua' }
```
Done

RunsFor commented on src/lua/init.lua:
```diff
@@ -86,81 +106,81 @@ local function load_lua(file)
-    local path = "./?."..soext
-    return package.searchpath(name, path)
-end
+local function merge_strings(paths, templates)
```
Made this function inline

Totktonada replied:
But don't deleted here.

RunsFor replied:
Fixed

RunsFor commented on src/lua/init.lua:
```diff
@@ -44,6 +44,26 @@ tarantool_exit(int);
+local package_approot
+
+local function approot(path)
+    if path == nil then
```
Yep, did it

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +111,85 @@ local function load_lua(file)
+        local searchpaths = {}
+
+        if #opts.templates == 0 then
+            return paths
```
AFAIU the function return one path to a file accessible for reading. Here you return a table of paths, don't try to open them for read. Also the code never executes actually: why do you want to handle empty templates case? I would ban it right in gen_search_func().

RunsFor replied:
You are right. I wrongly inlined function which did "merging" strings from these tables. Actually there was a regression in unit tests. Now it's fixed, and this code is executed when no templates provided to `get_search_func`

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +111,85 @@ local function load_lua(file)
-            return err
+
+        if #paths == 0 then
+            return opts.templates
```
The question for the block above mostly applicable here too. It is even more strange to return templates instead of a path. Also `paths` is never empty here, look how it is initialized.

RunsFor replied:
Thanks. Here was the same problem as in the previous comment. Also this code is dead, so I removed it

Totktonada commented on src/lua/init.lua:
```diff
@@ -86,81 +111,85 @@ local function load_lua(file)
-local function cwd_loader_func(lib)
-    local search_cwd = lib and search_cwd_lib or search_cwd_lua
-    local load_func = lib and load_lib or load_lua
+local function gen_search_func(opts)
```
Can you add an API comment here and for gen_loader_func()?

RunsFor replied:
Added comments for both functions

RunsFor replied:
Done

Totktonada commented on src/lua/init.lua:
```diff
@@ -169,22 +198,19 @@ local function rocks_loader_func(lib)
-        return package.searchpath(name, cpath and package.cpath or package.path)
-    end
-end
+local search_func_defs = {
```
`package.search()` did look into search_cwd_lib, search_rocks_lib and so on. Now it does not. Why?

RunsFor replied:
Didn't get you. There is no more `search_cwd_lib` and `search_rocks_lib` functions anymore. But the behaviour is actually the same (besides new "search root" functionality added)

RunsFor replied:
If you are saying about `search_cwd_lua` and `search_rocks_lua` then it was removed on purpose. See, `package.search` is used in Tarantool only for loading C functions (dlopen). I don't know why this function is in public interface or even why it was written in Lua space.

With that said when loading a C function, if coincidentally `package.search` finds a lua module, than following dlopen will fail. I made a simple test case when it may happen:

```
- error: "Failed to dynamically load module '': dlopen(/tmp/tntX4ScRS/function1.dylib,
    6): no suitable image found.  Did find:\n\t/tmp/tntX4ScRS/function1.dylib: file
    too short\n\t/Users/i.konyuhov/Workspace/tarantool/test/box/function1.lua: file
    too short"
```

In my opinion, `package.search` should not even look at the lua paths, but on the other hand, since `package.search` is a public function, users may have using it somehow (it is not in docs though). I tend to the first option, what do you thing?

Totktonada replied:
There is https://github.com/tarantool/avro-schema/commit/a8da39ab44fd007989f126592792655b50db2d69 (which however was reverted in https://github.com/tarantool/avro-schema/commit/e8faf0c1fbe0f700b21c02d15f3dbd37b2bec4a4). Here it is used for loading C library, so it is ok. Also I see that this function was mentioned twice in tarantool's public telegram chat.

I would not change the behaviour for now even despite that this looks buggy in some cases.

RunsFor replied:
Returned code, which is looking for lua dependencies as well as C dependencies

RunsFor replied:
Returned back to previous behaviour

Totktonada commented:
It is a kind of nitpicking, but while looking at the code I'm stuck several times with questions like 'whether we expect / return a semicolon-separated paths or just a single path here' and 'whether we expect only absolute path w/o `/../` path here or the path can be arbitrary'. Comments about contracts would be helpful IMHO, but I don't insist on this.

RunsFor commented on src/lua/init.lua:
```diff
@@ -44,6 +44,33 @@ tarantool_exit(int);
+        -- We need to get path to the function lower on the stack
+        path = debug.sourcedir(3)
+    elseif path == box.NULL then
+        path = ''
```
Agree. Unset `package_searchroot` var when `box.NULL` is passed

RunsFor commented on src/lua/init.lua:
```diff
@@ -44,6 +44,33 @@ tarantool_exit(int);
+
+local function setsearchroot(path)
+    if not path then
+        -- We need to get path to the function lower on the stack
```
✔️

RunsFor commented on src/lua/init.lua:
```diff
@@ -44,6 +44,33 @@ tarantool_exit(int);
+
+local function setsearchroot(path)
+    if not path then
+        -- We need to get path to the function lower on the stack
```
Fixed the wording

RunsFor commented on src/lua/init.lua:
```diff
@@ -44,6 +44,33 @@ tarantool_exit(int);
+    elseif path == box.NULL then
+        path = ''
+    else
+        assert(type(path) == 'string', 'Application directory must be a string')
```
Changed it to search root in code, tests and commit message

Totktonada commented:
I made several minor changes:

* Fit comments to 66 symbols, fit all code lines to 80 symbols.
* Got rid of the hack with package.{path,cpath} in path_fn().
* Made opts.templates non-optional, because of the previous point.

@RunsFor Can you look into the new code?

Totktonada commented:
@Gerold103 Can you look into this issue, please?

Gerold103 commented on src/lua/init.lua:
```diff
@@ -86,81 +111,75 @@ local function load_lua(file)
+--     multiple paths by expanding base path up to the root ('/')
+-- @return a searcher function which builds a path template and
+-- calls package.searchpath
+local function gen_search_func(opts)
```
Do you really need the options as a table? This is an internal function, so it is not necessary to pack arguments into a table for backward compatibility. Why not to make this function simpler?
```Lua
local function gen_search_func(path_fn, templates, need_traverse)
```
By the way, look at the new name of the flag - for flags we always add a verb like 'is_', 'need_', 'has_' etc.

RunsFor replied:
I usually prefer not to have too many function arguments. Also, named arguments are more readable from the caller point of view. But I don't insist. If this style is preferable in Tarantool code base, I'm ok to adjust.

RunsFor replied:
Switched to positional arguments

Gerold103 replied:
From the callers point of view - maybe. From the maintenance and function readability part it is not good. When a function takes only one 'opts' parameter, you need to spend much more time on documenting each option. The comment pads out. Additionally, you can accidentally misspell an opt name, and you won't notice it in code, the expected key will be just nil. I've faced that problem already.

Gerold103 commented on src/lua/init.lua:
```diff
@@ -86,81 +111,75 @@ local function load_lua(file)
+--     search function
+-- @return function a loader, which first search for the file and
+-- then loads it
+local function gen_loader_func(opts)
```
The comment as about `gen_search_func`.

RunsFor replied:
Switched to positional arguments here too

RunsFor commented:
@Gerold103 I corrected commit message as you pointed it out. Made docbot comment more detailed, but doesn't move all the commit message under it. I assume, for all the prerequisites one can go directly to the commit message anyways. Correct me if it is not appropriate.

I changed APIs of local search and loader functions generators to the positional arguments as well.

RunsFor commented:
This patch is also wanted in both `1.10` and `2.1` branches

RunsFor commented:
> @RunsFor Can you look into the new code?

@Totktonada Yes, I'm fine with that. Thanks!

Gerold103 commented:
> @Gerold103 I corrected commit message as you pointed it out. Made docbot comment more detailed, but doesn't move all the commit message under it. I assume, for all the prerequisites one can go directly to the commit message anyways. Correct me if it is not appropriate.
> 
> I changed APIs of local search and loader functions generators to the positional arguments as well.

Thanks for the fixes. The thing about commit message and doc request is that 1) before docs are pushed and a new Tarantool version is released, people will read tarantool/doc repo's issue; 2) on the contrary, people usually do not open commits, as you could think. At least Peter, in my experience, can sometimes decide not to read it when documents an issue, and it is something from appserver. Or probably he just omits part not included into a doc request because thinks of it not being important enough.
